src/Tmds.Ssh/KerberosCredential.cs

```diff
@@ -0,0 +1,339 @@
+    public KerberosCredential(NetworkCredential? credential = null, bool delegateCredential = false, string? serviceName = null)
+    {
+        if (!string.IsNullOrWhiteSpace(credential?.UserName) && string.IsNullOrWhiteSpace(credential?.Password))
+        {
+            throw new ArgumentException("credential Password cannot be null or an empty string.", nameof(credential));
+        }
+        _kerberosCredential = credential ?? CredentialCache.DefaultNetworkCredentials;
+        _delegateCredential = delegateCredential;
+        _serviceName = serviceName;
+    }
+
+    internal async Task<bool> TryAuthenticate(SshConnection connection, ILogger logger, SshClientSettings settings, SshConnectionInfo connectionInfo, CancellationToken ct)
+    {
+        // RFC uses hostbased SPN format "service@host" but Windows SSPI needs the service/host format.
+        // .NET converts this format to the hostbased format expected by GSSAPI for us.
+        string spn = string.IsNullOrEmpty(_serviceName) ? $"host/{connectionInfo.Host}" : _serviceName;
+
+        // The SSH messages must have a username value which maps to the target user we want to login as. We use the
+        // client supplied username as the target user. The Kerberos principal credential is only used for the
+        // authentication stage that happens next.
+        string userName = settings.UserName;
+        logger.AuthenticationMethodGssapiWithMic(userName, _kerberosCredential.UserName, spn, _delegateCredential);
+
+        bool isOidSuccess = await TryStageOid(connection, logger, userName, ct).ConfigureAwait(false);
+        if (!isOidSuccess)
+        {
+            return false;
+        }
+
+        var negotiateOptions = new NegotiateAuthenticationClientOptions()
+        {
+            AllowedImpersonationLevel = _delegateCredential ? TokenImpersonationLevel.Delegation : TokenImpersonationLevel.Impersonation,
+            Credential = _kerberosCredential,
+            Package = "Kerberos",
+            // While only Sign is needed we need to set EncryptAndSign for
+            // Windows client support. Sign only will pass in SECQOP_WRAP_NO_ENCRYPT
+            // to MakeSignature which fails.
+            // https://github.com/dotnet/runtime/issues/103461
+            RequiredProtectionLevel = ProtectionLevel.EncryptAndSign,
+            // While RFC states this should be set to "false", Win32-OpenSSH
+            // fails if it's not true. I'm unsure if openssh-portable on Linux
+            // will fail in the same way or not.
+            RequireMutualAuthentication = true,
+            TargetName = spn,
+        };
+
+        using var authContext = new NegotiateAuthentication(negotiateOptions);
+        bool isAuthSuccess = await TryStageAuthentication(connection, logger, authContext, ct).ConfigureAwait(false);
+        if (!isAuthSuccess)
+        {
+            return false;
+        }
+
+        try
+        {
+            using var message = authContext.IsSigned
```

Comment:
What determines whether `IsSigned` is `true`/false`?

Can you add some comment that explains why there are two cases here?

Reply:
It's based on the SSPI/GSSAPI flags in the exchange. We request both signing/encryption https://github.com/jborean93/Tmds.Ssh/blob/ed9e230bb60e3fca7deac7ba92caf2354c58c541/src/Tmds.Ssh/KerberosCredential.cs#L100 but the server can technically not honour this and not reply with the signing/encryption flags. While all implementations I know off will always sign for Kerberos I'm just following the spec to have both scenarios here. I'll add a comment.